MVP

belfortep opened the pull request:
## Summary
This PR is for the MVP of the `LC3 VirtualMachine`, all the instructions and traps are working, can test them with a `cargo test` also added the feature to run programs from `.obj` files with `cargo run -- -f FILE_NAME`
 
### TODO
- Add documentation

belfortep committed:
Can Add 2 numbers in same register

belfortep committed:
Can add values from different register to another register

belfortep committed:
Can AND different registers

belfortep committed:
Can negate the values of registers

belfortep committed:
Can branch the program

belfortep committed:
Load and Store to memory

belfortep committed:
Added load functions

belfortep committed:
Added jump functions and all of enum

belfortep committed:
Added store functions

belfortep committed:
Load and Store tested

belfortep committed:
Can jump to subroutines

belfortep committed:
Can store and load with base + offset

belfortep committed:
Can unconditionally jumps

belfortep committed:
Added traps functions

belfortep committed:
Added traps function to VM

belfortep committed:
Changed enums with struct constants

belfortep committed:
Changed to use enum with from for instructions opcode

belfortep committed:
Moved instructions from VM to own file

belfortep committed:
Moved trap handler functions to own file

belfortep committed:
Added from implementation for Register

belfortep committed:
Fix typo, programm -> program

belfortep committed:
Starting instructions refactor

belfortep committed:
VM parses instructions, instructions just executes them

belfortep committed:
Fix typo from tests

belfortep committed:
Changed tests numbers style for better readiblity

belfortep committed:
Changed function from public to private

belfortep committed:
Can receive keyboard input

belfortep committed:
small style fix

belfortep committed:
Moved register enum to own file

lazcanoluca commented on src/virtual_machine/instructions.rs:
```diff
@@ -0,0 +1,246 @@
+    source_one_register: Register,
+    source_two_register: Register,
+) {
+    let new_register_value = virtual_machine.read_register(source_one_register)
```
You'll have to use `.wrapping_add()`. If the sum of the contents of the registers overflow `u16`, the program will panic, which is not the desired behavior. This comment applies for all additions in operations.

belfortep replied:
Oh, didn't know about that, in another branch was casting from u32 and u16 as crazy because of that

lazcanoluca commented on src/virtual_machine/lc3_virtual_machine.rs:
```diff
@@ -0,0 +1,384 @@
+
+    pub fn process_input(&mut self, instruction: u16) {
+        let instruction_opcode = Instruction::from(instruction >> 12);
+        match instruction_opcode {
```
Is there a reason for not including the instruction construction directly in the corresponding functions? I believe it would be cleaner and easier to test.

juanimedone replied:
I definitely support this.

belfortep replied:
It was that way at first, but after thinking about it, I don't feel it makes sense that the instructions should have that responsability, as mentioned in another comment, the VM should decode and decide what instruction to execute, the instructions do the rest

lazcanoluca commented on src/virtual_machine/lc3_virtual_machine.rs:
```diff
@@ -0,0 +1,384 @@
+        self.memory[memory_address as usize] = value_to_write;
+    }
+
+    pub fn process_input(&mut self, instruction: u16) {
```
I don't think this function name is the best. What is "input"? Does the vm really _process an input_, rather than _decoding_ and _executing_ something that is already "inputted" (the loaded program)?

belfortep replied:
That's right, was the first name that came to my mind and seems very bad now that you mention it, definitely changing it

lazcanoluca commented on src/virtual_machine/lc3_virtual_machine.rs:
```diff
@@ -0,0 +1,384 @@
+    }
+}
+
+#[cfg(test)]
```
I think the instructions tests should be in the `instructions` module.

belfortep replied:
Oops, forgot about that, at first all the code was in this file and didn't notice

lazcanoluca commented on src/virtual_machine/lc3_virtual_machine.rs:
```diff
@@ -0,0 +1,384 @@
+
+use super::{instructions::*, register::Register, trap::Trap};
+
+const AMMOUNT_OF_REGISTERS: usize = 10;
```
I think this is a bit weird, it probably belongs in the `register` module. Having a second "source of truth" about the registers in an unrelated module is unexpected.

belfortep replied:
That's true, the creation of the `register` module was a last-second change and missed that one

lazcanoluca commented on src/virtual_machine/lc3_virtual_machine.rs:
```diff
@@ -0,0 +1,384 @@
+
+use super::{instructions::*, register::Register, trap::Trap};
+
+const AMMOUNT_OF_REGISTERS: usize = 10;
```
```suggestion
const AMOUNT_OF_REGISTERS: usize = 10;
```

lazcanoluca commented on src/virtual_machine/lc3_virtual_machine.rs:
```diff
@@ -0,0 +1,384 @@
+
+impl LC3VirtualMachine {
+    pub fn new(program_counter_start: u16) -> Self {
+        let mut registers = vec![0; AMMOUNT_OF_REGISTERS];
+        registers[Register::ProgramCounter as usize] = program_counter_start;
+        Self {
+            registers,
+            memory: vec![0; 1 << 16],
+        }
+    }
+
+    pub fn read_register(&self, source_register: Register) -> u16 {
+        self.registers[source_register as usize]
+    }
+
+    pub fn update_register(&mut self, destination_register: Register, new_register_value: u16) {
+        self.registers[destination_register as usize] = new_register_value;
+    }
```
I'm not convinced about this. If there is a `register` module, don't its operations belong to said module? Maybe use a `Registers` struct and delegate some responsability to it.

belfortep replied:
Yes, a `Registers` struct would make things clearer and more descriptive

lazcanoluca reviewed (commented):
Looks good overall. I left a few comments about things I found unexpected or would change, but let me know if you don't agree.

juanimedone commented on src/virtual_machine/lc3_virtual_machine.rs:
As a suggestion, you could name this file shorter. Something like `lc3.rs` or `vm.rs`.

belfortep replied:
In general short names makes me uncomfortable, I always forgot about the meaning of things, even in the same day 😅

juanimedone replied:
Ok it's fine then.

juanimedone commented on src/virtual_machine/lc3_virtual_machine.rs:
```diff
@@ -0,0 +1,384 @@
+    }
+
+    #[test]
+    fn can_add_two_numbers_in_differents_registers() {
```
```suggestion
    fn can_add_two_numbers_in_different_registers() {
```
A typo, same happens with some the following tests.

juanimedone reviewed (changes requested):
Apart from what @lazcanoluca mentioned, the rest of the code looks good to me. 

The thing is that for this to be a proper MVP I believe it should be able to run an `obj` file succesfully. You are close to it, just add the files provided [here](https://www.jmeiners.com/lc3-vm/#running-the-vm) (`2048.obj` and `rogue.obj`) and a few more lines of code to read them. It should work fine!

belfortep committed:
Fix typo in constant name

Co-authored-by: Luca Lazcano <64329820+lazcanoluca@users.noreply.github.com>

belfortep committed:
Fix typo in tests names

Co-authored-by: Juan Ignacio Medone Sabatini <jmedone@fi.uba.ar>

belfortep committed:
Changed sum operator for wrapping add

belfortep committed:
Merge branch 'mvp' of github.com:belfortep/lc3-vm into mvp

belfortep committed:
Changed process input function to a more descriptive name

belfortep committed:
Moved instructions test from VM to instruction file

belfortep committed:
Moved amount of registers constant to registers file

belfortep committed:
Created registers structure

belfortep committed:
Fixed typo in tests names

belfortep committed:
Added possibility to run programs from .obj files

belfortep commented:
Just added the feature to run programs from .obj files, you can try them with :
`cargo run -- -f ./object_files/rogue.obj`
or
`cargo run -- -f ./object_files/2048.obj`

And fixed the different comments 🤝

juanimedone reviewed (approved):
It works fine now! For the next iteration try disabling the input buffering to make the games more dynamic (so that the user doesn't need to press `Enter` after entering any key). [Here](https://www.jmeiners.com/lc3-vm/#platform-specifics) you have some info about that.

lazcanoluca commented:
It looks better now! The addition for loading `.obj` files makes it a true MVP. I would continue working on what Juani said, about the input buffering, so the game is dynamic. It can be it's own PR. Approving this one!

lazcanoluca reviewed (approved):
It looks better now! The addition for loading .obj files makes it a true MVP. I would continue working on what Juani said, about the input buffering, so the game is dynamic. It can be it's own PR. Approving this one!
